### ISSUE TITLE
py.test --pyargs my_pkg doesn't run pytest_addoption in my_pkg/conftest.py

### ISSUE THREAD

**Issue by Korijn** [pytest-dev/pytest#1596]
As part of our build process, we run the test suite _after_ creating the packaging to check that the installed package works as intended. We recently add an option to py.test for some conditional testing, by placing a function `pytest_addoption` in the `conftest.py` file that is in the root of the package (so next to `my_pkg/__init__.py`). When calling py.test on the source, this works as intended. Tests are skipped when we omit the option, and they are executed when we pass the option.

`my_pkg/conftest.py`:
```python
def pytest_addoption(parser):
    """Adds options to the pytest parser."""
    parser.addoption("--azure-storage-emulator",
                     action="store_true",
                     help="run tests that depend on the azure storage emulator")
```

However, when we run `py.test --pyargs my_pkg` in an environment where the built package is installed, we see the following error message (even though the test suite is run):

```
=================================== ERRORS ====================================
______________________________ ERROR collecting  ______________________________
..\..\envs\_test\lib\site-packages\my_pkg\storage\tests\test_azure.py:18: in <module>
    not pytest.config.getoption("--azure-storage-emulator"),
..\..\envs\_test\lib\site-packages\_pytest\config.py:1061: in getoption
    raise ValueError("no option named %r" % (name,))
E   ValueError: no option named 'azure_storage_emulator'
```

All other tests, including those depending on fixtures in the same `conftest.py` file, simply pass.

We're running on Windows and Linux, and both on dev machines as on CI, this scenario occurs. We're using the following related packages:

```
    pytest:                         2.9.1-py_1
    pytest-cov:                     2.2.1-py35_0
    pytest-mock:                    0.10.1-py35_0
    pytest-xdist:                   1.14-py35_0
    python:                         3.5.1-4
```

**Comment by nicoddemus** [comment]
Thanks for the report!

**Comment by nicoddemus** [comment]
By default `pytest_addoption` is not called on all `conftest.py` files, only at the root.

From [pytest_addoption](http://pytest.readthedocs.io/en/2.9.1/writing_plugins.html?highlight=pytest_addoption#_pytest.hookspec.pytest_addoption):

> Note
This function should be implemented only in plugins or conftest.py files situated at the tests root directory due to how py.test discovers plugins during startup.

I think this logic would have to be extended to also consider `conftest.py` files at the root of packages when using the `--pyargs`.

**Comment by Korijn** [comment]
Right, I thought that included the root of packages. Guess not. :)

**Comment by jessecarbon** [comment]
Just FYI, I found this workaround which works - https://stackoverflow.com/a/43747114/2149641

However it basically makes your tests a callable module instead of using `--pyargs`.

**Comment by larsks** [comment]
I showed up here to file exactly this bug but I see someone has beat me to it.  I want to be able to execute tests from an installed location, where my "root" `conftest.py` is located in the directory corresponding to `mypackage.tests`.  I was hoping that `pytest --pyargs mypackage.tests` would locate it correctly.

**Comment by bashtage** [comment]
Any progress on fixing this issue?  It has recently cropped up in statsmodels.

**Comment by Juanlu001** [comment]
Also bitten by this while running the statsmodels tests... @al1190

**Comment by larsks** [comment]
Hey, hello 2017 larsks! I just into the same problem **again** and forgot that I had previously filed this bug. @nicoddemus, is this a particularly intractable problem, or just not an interesting one?

**Comment by nicoddemus** [comment]
Hey @larsks,

Now that the dust has settled a bit with pytest 5.0, I will see if I can get a shot at this. Thanks for the ping. 👍

**Comment by vxgmichel** [comment]
Hi all! I just ran into the same issue, does anyone know a work around? 

I couldn't find a way to run the tests within an installed package with custom options, apart from:
  - pointing pytest to the test files using the full path:
    ```shell
    $ pytest myenv/lib/python3.7/site-packages/mypackage/tests --custom-option foo
    ```
  - move `pytest_addoption` to a separate `plugin` module and provide it to pytest through `-p`:
    ```shell
    $ pytest -p mypackage.tests.plugin --pyargs mypackage.tests --custom-option foo
    ```
  - move `pytest_addoption` to a separate `plugin` module and register it using setuptools (it isn't great because this specific module gets loaded at every pytest invocations but it only makes sense for the tests in this particular package)

Oh, and thank you for your work :)

**Comment by bashtage** [comment]
I use a pattern where I define a test function in my projects `__init__` that looks like:

```
def test(
    extra_args: Optional[Union[str, List[str]]] = None,
    exit: bool = True,
    append: bool = True,
):
    import sys

    try:
        import pytest
    except ImportError:
        raise ImportError("Need pytest to run tests")

    cmd = ["--tb=auto"]
    if extra_args:
        if not isinstance(extra_args, list):
            pytest_args = [extra_args]
        else:
            pytest_args = extra_args
        if append:
            cmd += pytest_args[:]
        else:
            cmd = pytest_args
    pkg = os.path.dirname(__file__)
    cmd = [pkg] + cmd
    print("running: pytest {}".format(" ".join(cmd)))
    status = pytest.main(cmd)
    if exit:
        sys.exit(status)

```


This lets me test from an install with

`python -c "import pkg;pkg.test()` or `python -c "import pkg;pkg.test(['-r a','--skip-slow','-n auto',...)`

**Comment by vxgmichel** [comment]
@bashtage Oh thanks, I didn't notice a [similar workaround](https://stackoverflow.com/a/43747114/2149641) had [already been shared](https://github.com/pytest-dev/pytest/issues/1596#issuecomment-315069111) earlier in the thread. 

Sorry for the noise :sweat_smile:

**Comment by max-sixty** [comment]
How feasible would it be to allow defining options in a `conftest.py` within the tests path? 

That would mean we didn't need to put the `conftest.py` file in the root, and so it wouldn't get missed by python's packaging (I think).

**Comment by yarikoptic** [comment]
> Now that the dust has settled a bit with pytest 5.0, I will see if I can get a shot at this. Thanks for the ping.

We stumbled upon this issue in dandi project as well.  Please let us know if dust conditions improved.  Cheers.

**Comment by pelson** [comment]
I did a little bit of digging here, and believe that the bug is https://github.com/pytest-dev/pytest/blob/103b2b67b015b5f677eef471c3d34706b5efedaa/src/_pytest/config/__init__.py#L570

In the case when `pyargs is True`, we should not treat the argument as a Path, and rather we need to resolve the name.


I quickly hacked together a fix:

```diff
diff --git a/src/_pytest/config/__init__.py b/src/_pytest/config/__init__.py
index 56b047196..24d0c1c53 100644
--- a/src/_pytest/config/__init__.py
+++ b/src/_pytest/config/__init__.py
@@ -566,8 +568,18 @@ class PytestPluginManager(PluginManager):
         self._noconftest = noconftest
         self._using_pyargs = pyargs
         foundanchor = False
         for initial_path in args:
-            path = str(initial_path)
+            if pyargs:
+                import importlib.util
+                # Note: some more care needed... might be that we specified a submodule, and we should use its directory.
+                mod_spec = importlib.util.find_spec(initial_path)
+                # Note: Some more care needed... might be not a path.
+                path = str(pathlib.Path(mod_spec.origin).parent)
+            else:
+                path = str(initial_path)
             # remove node-id syntax
             i = path.find("::")
             if i != -1:
```

There is still a bit of work to determine the precise behaviour (e.g. in the case of submodules, and non directory-based modules), but it is already a good start (it fixed the problem for me).
